sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,311 @@
+  ptrdiff_t offset = plic_offset_from_reg0(irq);
+  ptrdiff_t ie_offset;
+  switch (plic->target) {
+    case kDifPlicIbex0:
+      ie_offset = RV_PLIC_IE00 + offset;
+      break;
+    default:
+      return false;
+  };
+
+  uint8_t bit_index = plic_bit_index_from_irq_id(irq);
+  if (enable == kDifPlicEnable) {
+    reg32_nonatomic_set_bit(plic->base_addr, ie_offset, bit_index);
+  } else {
+    reg32_nonatomic_clear_bit(plic->base_addr, ie_offset, bit_index);
+  }
+
+  return true;
+}
+
+bool dif_plic_irq_le_control(const dif_plic_t *plic, dif_plic_irq_src_id_t irq,
+                             dif_plic_on_off_t enable) {
+  if (plic == NULL || !plic_irq_source_validate(irq) ||
+      !plic_on_off_valid(enable)) {
+    return false;
+  }
+
+  uint8_t bit_index = plic_bit_index_from_irq_id(irq);
+  ptrdiff_t offset = plic_offset_from_reg0(irq);
+  ptrdiff_t le_offset = RV_PLIC_LE0 + offset;
+  if (enable == kDifPlicEnable) {
+    reg32_nonatomic_set_bit(plic->base_addr, le_offset, bit_index);
+  } else {
+    reg32_nonatomic_clear_bit(plic->base_addr, le_offset, bit_index);
+  }
+
+  return true;
+}
+
+bool dif_plic_irq_priority_set(const dif_plic_t *plic,
+                               dif_plic_irq_src_id_t irq, uint32_t priority) {
+  if (plic == NULL || !plic_irq_source_validate(irq) ||
+      !plic_validate_priority(priority)) {
+    return false;
+  }
+
+  ptrdiff_t offset = plic_priority_offset_from_reg0(irq);
+  ptrdiff_t prio_offset = RV_PLIC_PRIO0 + offset;
+  reg32_write(plic->base_addr, prio_offset, priority);
+
+  return true;
+}
+
+bool dif_plic_target_threshold_set(const dif_plic_t *plic, uint32_t priority) {
+  if (plic == NULL || !plic_validate_priority(priority)) {
+    return false;
+  }
+
+  // Write to the target specific Threshold register
+  switch (plic->target) {
+    case kDifPlicIbex0:
+      reg32_write(plic->base_addr, RV_PLIC_THRESHOLD0, priority);
+      break;
+    default:
+      return false;
+  };
+
+  return true;
+}
+
+bool dif_plic_irq_ip_status(const dif_plic_t *plic, dif_plic_irq_src_id_t irq,
+                            bool *set) {
+  if (plic == NULL || set == NULL || !plic_irq_source_validate(irq)) {
+    return false;
+  }
+
+  uint8_t bit_index = plic_bit_index_from_irq_id(irq);
+  ptrdiff_t offset = plic_offset_from_reg0(irq);
+  ptrdiff_t ip_offset = RV_PLIC_IP0 + offset;
+  *set = reg32_get_bit(plic->base_addr, ip_offset, bit_index);
+
+  return true;
+}
+
+bool dif_plic_claim_irq(dif_plic_t *plic) {
+  if (plic == NULL || plic_irq_claimed(plic)) {
+    return false;
+  }
+
+  // Get the IRQ ID from the target specific CC register
+  uint32_t irq_id;
+  switch (plic->target) {
+    case kDifPlicIbex0:
+      irq_id = reg32_read(plic->base_addr, RV_PLIC_CC0);
+      break;
+    default:
+      return false;
+  };
+
+  // Validate the IRQ source, and determine which peripheral it belongs to
+  dif_plic_irq_src_id_t irq = (dif_plic_irq_src_id_t)irq_id;
+  if (irq == kDifPlicNoIrq) {
```

Comment:
I think that this big if chain would be much more readable as a giant switch over the entire enum, even if you wind up writing thirty fallthrough cases in a row for GPIO.

Reply:
How about an array of structs that store ranges for each peripheral? This block of code would be reduce to a loop.